storage/inkless/src/main/java/io/aiven/inkless/cache/CaffeineCache.java

```diff
@@ -42,20 +61,30 @@ public void close() throws IOException {
     }
 
     @Override
-    public FileExtent computeIfAbsent(final CacheKey key, final Function<CacheKey, FileExtent> mappingFunction) {
-        final CompletableFuture<FileExtent> future = new CompletableFuture<>();
-        final CompletableFuture<FileExtent> existingFuture = cache.asMap().computeIfAbsent(key, (cacheKey) -> {
-            return future;
+    public CompletableFuture<FileExtent> computeIfAbsent(
+        final CacheKey key,
+        final Function<CacheKey, FileExtent> load,
+        final Executor loadExecutor
+    ) {
+        // Caffeine's AsyncCache.get() provides atomic cache population per key.
+        // When multiple threads concurrently request the same uncached key, the mapping function
+        // is invoked only once, and all waiting threads receive the same CompletableFuture.
+        // This guarantees that the load function is called exactly once per key, preventing duplicate
+        // fetch operations from object storage.
+        return cache.get(key, (k, defaultExecutor) -> {
+            // Use the provided executor instead of Caffeine's default executor.
+            // This allows us to control which thread pool handles the fetch and blocks there,
+            // while Caffeine's internal threads remain unblocked, so cache operations can continue to be served.
+            return CompletableFuture.supplyAsync(() -> load.apply(k), loadExecutor)
+                .whenComplete((result, throwable) -> {
+                    // Evict the entry if the future completed exceptionally.
+                    // While Caffeine has built-in failed future cleanup, it happens asynchronously.
+                    // Explicit invalidation ensures immediate removal for faster retry on subsequent requests.
+                    if (throwable != null) {
+                        cache.synchronous().invalidate(key);
```

Comment:
nit: Should it be `k` instead of `key`? Using `k` I think we will guarantee that even for some reason the key will be mutated then we will properly invalidate the entry.